### DISCUSSION TITLE
Responding to Request for Collaborators

### DISCUSSION THREAD

**Discussion by JustinTimperio** [brittonhayes/pillager#30] [General]
Hi @brittonhayes,

I saw that you are currently looking for collaborators on this project and I have some interest in working on this. I recently built a tool (sadly a little too small for a repo) that could rapidly scan an entire disk's file tree into a compressed JSON object for exfiltration by an attacker. Rather than apply scanning rules on the client end, the object was to retrieve a list of all files as fast as possible, then let the attacker use their CPU resources to analyze the file paths, then return for the interesting files at a later point.

I'm not sure this work is directly relevant but it may provide you with some performance optimizations. I used (godirwalk)[https://github.com/karrick/godirwalk] which you may find useful. In benchmarks, it destroys pretty much everything else :) 

Feel free to DM at:
- @jtimperio:matrix.org
- justintimperio@gmail.com

**Comment by brittonhayes** [comment]
Hey @JustinTimperio! I appreciate you reaching out boss. I'm totally open to contributions or even just chatting out ideas to make this tool useful and performant for many different scenarios.

I think in the spirit of openness, if you ever have a new idea or concept for the repository, I'd love to try and keep stuff as public GitHub discussion posts as much as possible. Gives more people an opportunity to participate in the discussion and keeps the evolution of the project very transparent :smile:

As far as your idea goes, I think that's a great concept. For context, as of right now Pillager is backed entirely by Gitleaks' Non-git scanner so it functions much in the same way by just jumping straight into the files and digging for matches. I think there's a potential for a really nice hybrid or even just joining between your idea and the way it currently functions. The pre-scan approach almost sounds like it could be executed as a separate command, e.g. `pillager gather` - returning that compressed JSON object of the file tree. Then allow the user to follow up on that command with `pillager hunt -f my_file_tree.json`

**Reply by JustinTimperio** [reply]
Hi @brittonhayes,

Sure, I am still getting used to developing security tools in the open :sweat_smile: 

Yes, I think a pre-scanner solution would be optimal here for a number of reasons but mainly it would reduce the amount of pre-existing code that would need to be changed. I have read the code base lightly and while you are using a lot of really interesting go features and packages, it has obfuscated the control flow a little. I suspect I just need to put more time in the codebase and get my feet wet.

So a consideration here is resource usage. The reason my little tool is writing to a compressed buffer is because the size of the returned file list ballons to over 1GB of memory on machines with just 30GB of storage. The trade-off though is that at some point you need to convert it back to something usable with CPU cycles then also parse a massive list of entries with more cycles. 

I'm not sure what your runtime target or use case is though which may changes things. Do you see this as a Red Team or Blue Team tool? 

From my perspective that is basically the core design consideration. If this is a tool for attackers then runtime, resource usage, dependency management, and "embeddable-ness" become far more pressing concerns. As the defender, far more resources can be allocated to ensuring potential leaks have been closed with way less consideration for performance.

Best,
Justin

**Reply by JustinTimperio** [reply]
Just threw it into a repo so you can look at how it functions. It's a very tiny repo which I think is pretty straightforward. https://github.com/Project-Lernaean/goDiskDump

**Reply by brittonhayes** [reply]
> I have read the code base lightly and while you are using a lot of really interesting go features and packages, it has obfuscated the control flow a little. 

There are 1000% some piece of the hunter pkg specifically that I think need some love. Right now it's got some pretty clear moments of "shiny ball syndrome" where I've added a package without heavily considering the benefits of it's addition. I think the best example here is the afero.memfs. 

Additionally, I think that the [hound](https://github.com/brittonhayes/pillager/blob/33365ee4cd4eaa06cbca0bdb04400a8f1d5b8ddc/hunter/hound.go#L14-L73), [templates](https://github.com/brittonhayes/pillager/blob/main/hunter/template.go), and [output formats](https://github.com/brittonhayes/pillager/blob/33365ee4cd4eaa06cbca0bdb04400a8f1d5b8ddc/hunter/format.go#L5-L40) should be moved over into their own package. The functionality of printing the results is uncomfortably coupled with the scanning and I think that leads to some unnecessary complexity in my codebase.

> So a consideration here is resource usage. The reason my...

Are you saying that the file list returned is 1GB when uncompressed or compressed? And if uncompressed, what is the resource allocation improvement you're seeing when writing to the compressed buffer? 

edit: Just saw your repo link. taking a peek.

> I'm not sure what your runtime target or use case is though which may changes things. Do you see this as a Red Team or Blue Team tool?

That's a damn good question and I really hadn't come to terms with the fact that it's probably important to decide this. I am really leaning towards red team tool. For the primary reason that Gitleaks does a phenomenal job at handling the Blue Team use-case. So Pillager has an opportunity to be turned into something very portable and flexible. Rather than something used as a git pre-commit checker, I see pillager as a tool belt addition for security researchers. The primary goals in that case IMO would be speed of execution, cross-platform compatibility, simple command line interface/experience, and ease of reporting findings.

I appreciate you asking this. Hadn't faced that fork in the road yet but being directly asked makes it much easier to pick a direction.

**Comment by JustinTimperio** [comment]
> Right now it's got some pretty clear moments of "shiny ball syndrome" where I've added a package without heavily considering the benefits of it's addition.

I think this can be very difficult, as Go has a very active community with some amazing projects. I think the core team tho does a good job of following the KISS principle. 

> Are you saying that the file list returned is 1GB when uncompressed or compressed?

It ends up around 1GB _uncompressed_ in mem. I think a lot of this is just poor go garbage collection. In any case, the compression brings it down into the 100-400MB range depending on the file system.

> edit: Just saw your repo link. taking a peek.

It was a 1-hour hack job so it's not very clean <insert yikes here>

>  I am really leaning towards red team tool. For the primary reason that Gitleaks does a phenomenal job at handling the Blue Team use-case. 

I think this is a fair assessment. I also assumed from the name that as a "pillager" the goal would be to raid a file system for pretty much anything of value. These days I am primarily a grey hat hacker and I rarely run into git repos in the wild. As an attacker, I am interested in far more generic and commonplace file types. For example, personal info such as pictures, accounting records, and other background info can be used to build extremely powerful social engineering attacks. These file types are much harder to sort through and require a far more detailed inventory and scanning methodology than say a gitleaks.

**Reply by brittonhayes** [reply]
Okay so I dug around the the repo and I think you've got an idea worth exploring here. It seems like a very lightweight addition. I've made some tiny tweaks in this Github gist to just get a rough take on testing it out in docker. I also switched the GetPaths from `[]string` to `string` since I figured this will only ever be run on one OS at a time. I also removed the JSON Marshal Indent and changed the module to point at the Github repo. But that's about it. 

This runs and compiles wicked fast and seems like a very efficient way to extract directory paths for further parsing later. I'd almost say skip the json marshal until the end too. That'd allow it to be run at the end rather than on each new path. But I'm just spit balling here and I know this is like a PoC idea. 

TLDR, your idea looks great and I welcome any contributions you'd like to make to expand the functionality of pillager.

https://gist.github.com/brittonhayes/23b99941cafcfde0bbf9d7f717041fac

**Reply by JustinTimperio** [reply]
> I also switched the GetPaths from `[]string` to `string`

So this was actually done because Windows pathing sucks and doesn't mount drives in a single file tree. So a Windows device could have `C:\ D:\ F:\` but they would all be separate trees. To make that work you need to loop over every base path but i still haven't figured out how to accurately get a list of online drives.

> I'd almost say skip the json marshal until the end too.

The issue is that `buff` is being fed directly into the compressed zbuffer. Since the raw string is being converted directly to a compressed bytes array on ingest, the bytes would need to be decompressed, marshaled, then recompressed if that approach was taken. Basically, it works like `Read path --> Get Data --> Marshal --> Compress --> Write to results buff --> continue`

> your idea looks great and I welcome any contributions you'd like to make

Do you have any thoughts on building your own "gitleaks" style repo? Rather than specializing in a few edge case data exfill scenarios, turn pillager into its own rule-based system, where the library "pilages" the system for possible interesting files then presents them in a tiered incident response style report. The Report would provide 

1. A high-level system overview including os info, system usage, actively used folders, common extensions, ect.
2. A tiered list of suspect files
    - Tier 1: The big OOF section. Hardcoded passwords, unprotected ssh keys, login tokens, banking info, medical records, ect. 
    - Tier 2: The usual suspects. Photos, text and word documents, program files and configurations
    - Tier 3: TLDR: Basically files that are too large or complex to be properly categorized. These types of files are either useless or an actual gold mine.
    - Tier 4: The garbage bin: System files, precompiled binaries, ect
    
This is actually was what I was imagining for the `hunter` subpackage. Basically, pillager would act as a controller for a group of hunter subroutines that would raid the system for files. Each hunter would be applying its own analysis subroutines to extract and determine metadata about the files for the final report

**Reply by brittonhayes** [reply]
> I also switched the GetPaths from []string to string

OH! Okay I'm on the same page now. I fully misinterpreted that function. Makes sense.

> Read path --> Get Data --> Marshal --> Compress --> Write to results buff --> continue

Fair point. I didn't actually consider the re-doing the compression side of it. I see what you're doin now.

> Rather than specializing in a few edge case data exfill scenarios, turn pillager into its own rule-based system, where the library "pilages" the system for possible interesting files then presents them in a tiered incident response style report. The Report would provide...

I think we're definitely on the same page here. Where pillager would be the gatherer and presenter of the information and actions taken after that can be determined by the end user. Also lol at big oof section. 10/10 name. I also really dig the idea of breaking it into categories based on their severity for easy post-mortem analysis and filtering.

**Reply by JustinTimperio** [reply]
>  also really dig the idea of breaking it into categories based on their severity for easy post-mortem analysis and filtering.

I also like this because it would provide an API for what could be called a "static file system assessment tool". In the same way, we feed our code through static code analysis tools, "Pillager" could be made into a very compelling security assessment tool for both Red and Blud Team members. I also work on a project called ToRat (a red team tool) and adding the ability to call an external library to manage the analysis of an infected machine would be an amazing addition. From the other side, blue team members would be able to set up automated pipelines for collecting and logging potential holes created by misconfigured services.

This does raise a question for me though. How comfortable are you with rebasing/large commits? From the rough concept I have spec'ed in my head, it would be a fairly serious departure from your current codebase. 

It would involve:
- A fairly complex asynchronous control surface
- An entirely custom content flagging system
- A tiered file analysis tool (tiered hunters for different things all with self-managing attributes)
- A smart routine dispatcher capable of managing the load (unbounded parallelism is pretty obvious as an attacker)
- An API for integration with other tools 
- A reasonable report interface for admins to review results as blue team.

**Reply by brittonhayes** [reply]
I'm actually completely in favor of this. At the moment pillager feels very half-baked. The extent of its feature set is almost entirely in the customizable template output which isn't inherently unique. 

I think that large sweeping changes to the tool and repository would be for the better of the project and be the most likely way to turn this into something with more character and utility.

The only things I'd like to try to really stick to are:

- [The effective Go guidelines](https://golang.org/doc/effective_go)

- A simple command hierarchy. e.g. I'd like to keep the interface on the side of Docker or kubectl vs something like Hashcat where it can be overwhelming for new users.

- As much as possible, try to make PRs small and easily reviewable. It takes longer but since this is a side project outside of work, the more digestible a pull request is the more likely it's gonna get the love it needs.

- My current hope is to keep pillager on the neutral side of the security ethics line. So commands like torrat's shred command or similar, I'd hope to not do with this project. It should be purely for (like you said) file analysis, content flagging, and more things of that nature. Tldr: the information gathering stage of the penetration testing process.

Nothing different than really any other Go project but just putting those out as things that feel important to sustainable expansion of pillagers functionality. So all that being said, to answer your question yes. I am up for large overhauls of the code base given that the above bullets are kept in mind along the way 😊

**Reply by JustinTimperio** [reply]
I am in firm agreement on almost all these points. I stick to the Go guidelines pretty heavily. You'll also probably notice I am primarily a functional programmer and a huge nerd for KISS. Those are really my core axioms when writing software. 

>  So commands like torrat's shred command or similar, I'd hope to not do with this project.

I actually had never considered that destructive (its just a pure go version of the unix util) but I see your point. I think this is something that we all grapple with. There are a number of features I have kept out of ToRat, mainly features that would make it trivial to control a botnet. Sadly though I think at some point you will have to face the fact that Pillager, as a project, will result in abuse by bad actors. It's pretty likely script kiddies and other bad actors will most likely benefit from our work here. 

I think it is important though to keep in mind that the good these tools do in providing transparency to administrators and users far outweighs the cost. While this tool may be used by some kids to steal their friend's discord token, the administrators that add this as a risk detection pipeline will provide their users with substantially better security. 


So I will probably spin up some discussions here so I can keep the docs and discussions organized for us and others in the future but my current thought process is the following:

- The template output is fairly helpful actually. The backend should be able to be swapped out without a huge amount of effort.
- The core should be pretty fast to frame and build. The key will be to build a highly reliable and performant frame we can extend on. Then as others join or make advances in filtering and detection, updates can easily be applied to existing rule sets and sub-routines.
- I think most of the work here is the actual analysis of the files. (IE in the chain of `identity --> analyze --> respond`, most of the work is the middle stage. I suspect like 60% of the repo will be the hunter subroutines

**Reply by brittonhayes** [reply]
Man, you are very good at addressing any concerns I have. You explained that beautifully. I'm totally with you on it and like the way you're going with this. Please feel free to make discussion posts and let me know if you'd like me to add a kanban to the projects tab as well.

This is a new side of the coin for me, having been exclusively working on the blue team side of things so if at any point my replies come through as reserved, it's not a perspective of disagreement so much as it is one of my unfamiliarity. Always willing to hear out the angle, so please feel free to share any thoughts or differences of opinion as this goes along.

I'll set up a contributors.md and append the README to point people towards the discussions section in case of ideas.

**Reply by JustinTimperio** [reply]
> This is a new side of the coin for me, having been exclusively working on the blue team side of things so if at any point my replies come through as reserved, it's not a perspective of disagreement so much as it is one of my unfamiliarity.

This is totally justified. I have mostly lived in the grey areas of the industry, so I have had some time to deal with the ethics of these choices. I have built some tools I honestly never plan to release. On the flip side, by not releasing them, I leave potential holes in the applications and software I rely on. I have always found it a funny conundrum ethically. You can't make security good unless you break it, so it's always this race to break it fast and patch it before it becomes an issue. My opinion these days is that, for the most part, it's almost always worth publishing security vulnerabilities and hacking tools. 

My core caveat is that I don't cater to script kiddies. I see a lot of people make the mistake of trying to cater to this audience because it garners stars and usage. I personally see this as a negative. While making software simple is always a good thing, creating tools that specifically allow inexperienced and ethically dubious indiviudals to benefit from academic work is kind of a waste of good work. I have found this to be a really weird line to walk but thankfully I think "pillager" as a concept posses little ethical dilemmas. If someone is already inside your FS, you have a bigger issue, that has nothing to do with the stuff you leave unsecured on the system.
